control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+
+	// Determine Consul service selector values from pod names.
+	podPrefixes := make(map[string]any)
+	podExactNames := make(map[string]any)
+	for _, subset := range endpoints.Subsets {
+		for address := range allAddresses(subset) {
+			if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+				podName := types.NamespacedName{Name: address.TargetRef.Name, Namespace: req.Namespace}
+
+				// Accumulate owner prefixes and exact pod names for Consul workload selector.
+				// If this pod is already covered by a known owner prefix, skip it.
+				// If not, fetch the owner. If the owner has a unique prefix, add it to known prefixes.
+				// If not, add the pod name to exact name matches.
+				maybePodOwnerPrefix := getOwnerPrefixFromPodName(podName.Name)
+				if _, ok := podPrefixes[maybePodOwnerPrefix]; !ok {
+					var pod corev1.Pod
+					if err = r.Client.Get(ctx, podName, &pod); err != nil {
+						r.Log.Error(err, "failed to get pod", "name", podName.Name, "ns", req.Namespace)
+						errs = multierror.Append(errs, err)
+						continue
+					}
+					// Add to workload selector values.
+					// Pods can appear more than once in Endpoints subsets, so we use a set for exact names as well.
+					if prefix := getOwnerPrefixFromPod(&pod); prefix != "" {
+						podPrefixes[prefix] = true
+					} else {
+						podExactNames[podName.Name] = true
+					}
+				}
+			}
+		}
+	}
+
+	//TODO(zalimeni) Do we need to check service-enable label here on service/deployments/replicasets?
+	// Should we expect owner is annotated, and if so, is that consistent for non-replicaset workloads?
+	// If so, update to register and deregister conditionally.
+	if err = r.registerService(ctx, resourceClient, service, podPrefixes, podExactNames); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+
+	return ctrl.Result{}, errs
+}
+
+// allAddresses combines all Endpoints subset addresses to a single set. Service registration by this controller
+// operates independent of health, so we combine them here to simplify iteration.
+func allAddresses(addresses corev1.EndpointSubset) map[corev1.EndpointAddress]any {
+	m := make(map[corev1.EndpointAddress]any)
+	for _, readyAddress := range addresses.Addresses {
+		m[readyAddress] = true
+	}
+
+	for _, notReadyAddress := range addresses.NotReadyAddresses {
+		m[notReadyAddress] = true
+	}
+
+	return m
+}
+
+// getOwnerPrefixFromPodName extracts the owner name prefix from a pod name.
+func getOwnerPrefixFromPodName(podName string) string {
+	podNameParts := strings.Split(podName, "-")
+	return strings.Join(podNameParts[:len(podNameParts)-1], "-")
+}
+
+// getOwnerPrefixFromPod returns the common name prefix of the pod, if the pod is a member of a set with a unique name
+// prefix. Currently, this only applies to ReplicaSets.
+//
+// We have to fetch the owner and check its type because pod names cannot be disambiguated from pod owner names due to
+// the `-` delimiter and unique ID parts also being valid name components.
```

Comment:
Feel free to correct me if anyone knows better here. 🧵 for context [here](https://hashicorp.slack.com/archives/C04MG2UT3PA/p1692203149392719?thread_ts=1692199919.770389&cid=C04MG2UT3PA).